src/frequenz/quantities/marshmallow.py

```diff
@@ -0,0 +1,192 @@
+# License: All rights reserved
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+"""Custom marshmallow fields and schema."""
+
+from typing import Any, Type
+
+from marshmallow import EXCLUDE, Schema, ValidationError, fields
+
+from ._current import Current
+from ._energy import Energy
+from ._frequency import Frequency
+from ._percentage import Percentage
+from ._power import Power
+from ._quantity import Quantity
+from ._temperature import Temperature
+from ._voltage import Voltage
+
+
+class _QuantityField(fields.Field):
+    """Custom field for Quantity objects supporting per-field serialization configuration."""
+
+    def __init__(self, field_type: Type[Quantity], **kwargs: Any) -> None:
```

Comment:
I would specify the options explicitly here (like `serialize_as_string_default`), and only leave the `**kwargs` for the options that are forwarded to the `super()` if possible. Otherwise is hard to know which options it really takes in the docs (for the parent, you can always go check the parent class docs).

Reply:
That seems.. complicated?
For one, that field class and init c'tor is pretty much never used directly.

Why not keep the configuration as it is done here, through the "metadata" key
where we also configure all the other field specific things like validation, key name etc.. 

Like, I see zero reason to do this one thing differently than all the other things we can configure per-field?

Reply:
I guess I just don't understand this code very well. I thought this was a base class, but now I see it takes the `field_type` as argument instead. I don't know, compared to the previous version where there was a class for each field, for me this is **extremely** hard to read and understand, that's my main complain here :worried:

Reply:
It took me a while as well :sweat_smile:  

This class works for all quantities.
The only thing derived classes need to do is pass their specific quantity type.

Reply:
Deserialization automatically detects the type (float/str), serialization uses the schema default which can be overridden by a field-specific key.

Reply:
I still don't understand why, specially now that boilerplate code like this can be automatically generated with a few tabs, we want to keep dynamic unreadable code instead of having some much more obvious as self-explanatory. I see zero advantages of this approach compared to the verbose previous version and a lot of drawbacks.

Reply:
> I still don't understand why

Why what? Why dynamic creation? Because that would be a different comment thread..

I am not strongly against it, I just see zero benefit as noone uses the resulting code directly..
And I doubt it makes things much clearer really.. just lots of noise

Reply:
From the end user PoV I'm fine with making all the fields "private" and not export them at all if we think nobody will use them. For me one or the other, but the current approach it's just exporting something that's unusable for anyone reading the docs and only adding confusion.

In terms of code, I really don't understand this code and would really prefer not to have it in the repository, again, specially when there is a super clear and easy alternative. There is no need for any black magic here, it should be very simple. The only reason I can see for it is to save a few keystrokes.

That said, I don't feel strong enough to keep investing time on it, I'm fine to merge as is as long as we don't export the fields for now, then we can see in a future PR to improve it to be able to export the fields if there is a need for it.

Reply:
I see I'm always late with my comments that end up in "I'm fine with merging as is" :stuck_out_tongue_closed_eyes: